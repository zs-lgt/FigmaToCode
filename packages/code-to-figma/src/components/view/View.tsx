import * as React from 'react';
import { Frame, Rectangle, StyleSheet } from '../..';
import { RectangleProps } from '../rectangle/Rectangle';
import { FrameNodeProps } from '../frame/Frame';

export type ViewProps = FrameNodeProps | RectangleProps;

const View: React.FC<ViewProps> = props => {
    const style = (props.style && StyleSheet.flatten(props.style)) || {};
    if (props.children) {
        return (
            <Frame
                {...props}
                style={[
                    { backgroundColor: 'transparent' },
                    (process.env.REACT_FIGMA_WEB_DEFAULTS_ENABLED &&
                        ((style as any).display === 'flex' || (style as any).display === 'inline-flex') && {
                            flexDirection: 'row'
                        }) ||
                        undefined,
                    (process.env.REACT_FIGMA_WEB_DEFAULTS_ENABLED && { alignItems: 'stretch' }) || undefined,
                    style
                ]}
            />
        );
    } else {
        return <Rectangle {...(props as RectangleProps)} />;
    }
};

export { View };
