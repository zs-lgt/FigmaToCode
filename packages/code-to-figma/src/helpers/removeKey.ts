export const removeKey = node => {
    if (node && node.key) {
        delete node.key;
    }
    if (node && node.children) {
        node.children.forEach(removeKey);
    }
    return node;
};
