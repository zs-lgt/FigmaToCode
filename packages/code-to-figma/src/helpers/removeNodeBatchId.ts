export const removeNodeBatchId = node => {
    if (node && node.pluginData) {
        delete node.pluginData.nodeBatchId;
    }
    if (node && node.children) {
        node.children.forEach(removeNodeBatchId);
    }
    return node;
};
