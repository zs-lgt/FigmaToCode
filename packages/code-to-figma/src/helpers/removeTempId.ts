export const removeTempId = node => {
    if (node && node.pluginData) {
        delete node.pluginData.reactId;
    }
    if (node && node.children) {
        node.children.forEach(removeTempId);
    }
    return node;
};
