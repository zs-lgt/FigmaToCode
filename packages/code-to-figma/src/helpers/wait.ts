import { act } from 'react-test-renderer';

export function wait(amount = 0) {
    return new Promise(resolve => setTimeout(resolve, amount));
}

// Use this in your test after mounting if you need just need to let the query finish without updating the wrapper
export async function actWait(amount = 0) {
    await act(async () => {
        await wait(amount);
    });
}
