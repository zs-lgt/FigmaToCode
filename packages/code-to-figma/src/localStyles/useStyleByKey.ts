import { api } from '../rpc';
import * as React from 'react';
import { CommonStyle } from '../types';

export const useStyleByKey = (styleKey: string): BaseStyle => {
    const [style, setStyle] = React.useState<BaseStyle>(null);

    React.useEffect(() => {
        if (styleKey) {
            api.importStyleByKeyAsync(styleKey).then(setStyle);
        }
    }, [styleKey]);

    return style;
};
