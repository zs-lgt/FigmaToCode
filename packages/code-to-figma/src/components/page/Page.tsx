import * as React from 'react';
import { BaseNodeProps, ChildrenProps, ExportProps, StyleOf } from '../../types';
import { useYogaLayout } from '../../hooks/useYogaLayout';
import { YogaStyleProperties } from '../../yoga/YogaStyleProperties';
import { $currentPageTempId, api } from '../../rpc';
import { map } from 'rxjs/operators';
import { OnLayoutHandlerProps, useOnLayoutHandler } from '../../hooks/useOnLayoutHandler';
import { InheritStyleProvider } from '../../hooks/useInheritStyle';
import {
    GeometryStyleProperties,
    transformGeometryStyleProperties
} from '../../styleTransformers/transformGeometryStyleProperties';
import { StyleSheet } from '../../helpers/StyleSheet';
import { useImageHash } from '../../hooks/useImageHash';

export interface PageProps extends BaseNodeProps, ChildrenProps, ExportProps, OnLayoutHandlerProps {
    style?: StyleOf<YogaStyleProperties & GeometryStyleProperties>;
    isCurrent?: boolean;
    onCurrentChange?: (isCurrent: boolean) => void;
    backgrounds?: ReadonlyArray<Paint>;
}

export const useCurrentPageChange = (
    nodeRef: { current?: any },
    callback?: ((isCurrent: boolean) => void) | void,
    _isCurrent?: boolean
) => {
    React.useEffect(() => {
        const instance = nodeRef.current;
        const subject = $currentPageTempId.pipe(map((message: any) => message === instance.reactId));

        const subscription = subject.subscribe(value => callback && callback(value));

        return () => subscription.unsubscribe();
    }, []);

    React.useEffect(() => {
        if (_isCurrent) {
            api.setCurrentPage(nodeRef.current);
        }
    }, [_isCurrent]);
};

const Page: React.FC<PageProps> = props => {
    const nodeRef = React.useRef();
    const { onCurrentChange, ...otherProps } = props;
    useCurrentPageChange(nodeRef, onCurrentChange, props.isCurrent);

    const yogaChildProps = useYogaLayout({ nodeRef, ...otherProps });
    useOnLayoutHandler(yogaChildProps, props);

    const style = StyleSheet.flatten(props.style);

    const imageHash = useImageHash(style.backgroundImage);

    const pageProps = {
        ...transformGeometryStyleProperties('backgrounds', style, imageHash),
        ...otherProps,
        style
    };

    return (
        <InheritStyleProvider style={props.style}>
            <page {...pageProps} {...yogaChildProps} innerRef={nodeRef} />
        </InheritStyleProvider>
    );
};

export { Page };
