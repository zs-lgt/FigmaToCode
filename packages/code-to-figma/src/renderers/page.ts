import { baseNodeMixin } from '../mixins/baseNodeMixin';
import { saveStyleMixin } from '../mixins/saveStyleMixin';
import { exportMixin } from '../mixins/exportMixin';
import { pageMixin } from '../mixins/pageMixin';
import { PageProps } from '../components/page/Page';

export const page = (node: PageNode) => (props: PageProps) => {
    const page = node || figma.createPage();

    saveStyleMixin(page)(props);
    baseNodeMixin(page)(props);
    exportMixin(page)(props);

    pageMixin(page)(props);

    return page;
};
