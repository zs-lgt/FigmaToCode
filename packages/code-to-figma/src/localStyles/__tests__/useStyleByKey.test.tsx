import * as React from 'react';
import { createFigma } from 'figma-api-stub';
import { render } from '../../renderer';
import { wait } from '../../helpers/wait';
import { Text } from '../../components/text/Text';
import { View } from '../../components/view/View';
import { Page } from '../../components/page/Page';
import { removeNodeBatchId } from '../../helpers/removeNodeBatchId';
import { removeTempId } from '../../helpers/removeTempId';

const removeMeta = node => {
    return removeKey(removeNodeBatchId(removeTempId(node)));
};

jest.mock('nanoid');
import { nanoid } from 'nanoid';
import { useStyleByKey } from '../useStyleByKey';
import { removeKey } from '../../helpers/removeKey';

describe('useStyleByKey', () => {
    beforeEach(() => {
        // @ts-ignore
        global.figma = createFigma({
            simulateErrors: true
        });
        // @ts-ignore
        global.figma.importStyleByKeyAsync = jest.fn().mockReturnValue({ id: 'testStyleId' });
    });

    it('useTextStyle correctly applied', async () => {
        // @ts-ignore
        nanoid.mockReturnValue('test');
        const ViewComponent = () => {
            const fillStyle = useStyleByKey('testStyleId');
            return (
                <View
                    style={{
                        fillStyleId: fillStyle && fillStyle.id
                    }}
                />
            );
        };
        const TextComponent = () => {
            const textStyle = useStyleByKey('testStyleId');
            return (
                <Text
                    style={{
                        textStyleId: textStyle && textStyle.id
                    }}
                />
            );
        };
        await render(
            <Page>
                <ViewComponent />
                <TextComponent />
            </Page>
        );
        await wait();
        // @ts-ignore
        expect(removeMeta(figma.root)).toMatchSnapshot();
    });
});
