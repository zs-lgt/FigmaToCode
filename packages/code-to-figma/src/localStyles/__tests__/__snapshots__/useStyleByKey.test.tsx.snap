// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useStyleByKey useTextStyle correctly applied 1`] = `
ChildrenMixinStub {
  "children": Array [
    ExportMixinStub {
      "children": Array [],
      "config": undefined,
      "id": "0:1",
      "parent": [Circular],
      "type": "PAGE",
    },
    ExportMixinStub {
      "_backgrounds": Array [
        Object {
          "blendMode": "NORMAL",
          "color": Object {
            "b": 0.9607843160629272,
            "g": 0.9607843160629272,
            "r": 0.9607843160629272,
          },
          "opacity": 1,
          "type": "SOLID",
          "visible": true,
        },
      ],
      "children": Array [
        GeometryMixinStub {
          "_fills": Array [],
          "blendMode": "NORMAL",
          "config": Object {
            "isWithoutTimeout": false,
            "simulateErrors": true,
          },
          "constraints": Object {
            "horizontal": "MIN",
            "vertical": "MIN",
          },
          "cornerRadius": 0,
          "cornerSmoothing": 0,
          "effects": Array [],
          "exportSettings": Array [],
          "fillStyleId": "testStyleId",
          "id": "1:2",
          "isMask": false,
          "layoutAlign": "INHERIT",
          "layoutGrow": 0,
          "locked": false,
          "opacity": 1,
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{\\"fillStyleId\\":\\"testStyleId\\"}",
          },
          "strokeAlign": "INSIDE",
          "strokeCap": "NONE",
          "strokeWeight": 0,
          "strokes": Array [],
          "type": "RECTANGLE",
          "visible": true,
        },
        GeometryMixinStub {
          "_characters": "",
          "_fills": Array [
            Object {
              "color": Object {
                "b": 0,
                "g": 0,
                "r": 0,
              },
              "opacity": 1,
              "type": "SOLID",
            },
          ],
          "_fontName": Object {
            "family": "Inter",
            "style": "Regular",
          },
          "_textAutoResize": "WIDTH_AND_HEIGHT",
          "autoRename": false,
          "blendMode": "NORMAL",
          "config": Object {
            "isWithoutTimeout": false,
            "simulateErrors": true,
          },
          "constraints": Object {
            "horizontal": "MIN",
            "vertical": "MIN",
          },
          "effects": Array [],
          "exportSettings": Array [],
          "fontSize": 12,
          "hyperlink": null,
          "id": "1:3",
          "isMask": false,
          "layoutAlign": "INHERIT",
          "layoutGrow": 0,
          "letterSpacing": Object {
            "unit": "PIXELS",
            "value": 0,
          },
          "lineHeight": Object {
            "unit": "AUTO",
          },
          "locked": false,
          "opacity": 1,
          "paragraphIndent": 0,
          "paragraphSpacing": 0,
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{\\"textStyleId\\":\\"testStyleId\\"}",
          },
          "strokeAlign": "INSIDE",
          "strokeCap": "NONE",
          "strokeWeight": 0,
          "strokes": Array [],
          "textAlignHorizontal": "LEFT",
          "textAlignVertical": "TOP",
          "textCase": "ORIGINAL",
          "textDecoration": "NONE",
          "textStyleId": "testStyleId",
          "type": "TEXT",
          "visible": true,
        },
      ],
      "config": Object {
        "isWithoutTimeout": false,
        "simulateErrors": true,
      },
      "exportSettings": Array [],
      "id": "1:1",
      "parent": [Circular],
      "pluginData": Object {
        "isReactFigmaNode": "true",
        "reactStyle": "{}",
      },
      "type": "PAGE",
    },
  ],
  "config": undefined,
  "id": "0:0",
  "type": "DOCUMENT",
}
`;
